aten/src/ATen/cuda/CUDAFuture.h

```diff
@@ -31,6 +31,8 @@ struct TORCH_CUDA_API CUDAFuture : at::ivalue::Future {
   }
 
   void postMarkCompletedHook(const at::IValue& value) override {
+    currentDevice_ = c10::cuda::current_device();
```

Comment:
hmm, why we are recording the device when the future is marked as completed, instead of remember the device when the callback was inserted (through `then`/`addCallback`)?

Reply:
I guess those are two different approaches to this. The "philosophy" I was following is this: `then()` and `addCallback()` are used to perform a computation after another one is complete. If we were dealing with with sync operations, one would do this:
```
do_sth_sync()
do_sth_later()
```
but with async ops this needs to change and become
```
fut = do_sth_async()
fut.then(do_sth_later)
```
In the sync scenario, `do_sth_later()` runs in the same "environment"/"context" as `do_sth_sync()` (same current device, same current streams, ...). In this diff I was trying to recreate this in the async case, by "recording" the environment's state at the end of the async operation, and then "recreating" it in the callback.

Note that the approach we have for streams is _somewhat_ similar. We do _not_ run the callback in the streams that were current when the callback was inserted. (In honesty, we also do not run it in the streams that were current when the async op finished (as that would mean running computations in the I/O streams), but we do synchronize the "fresh" streams with those streams).

Note also that I am not opposed to changing this behavior, and replicate the "environment" that was current when the user inserted the callback. That has its own sets of advantages. For streams, for example, it allows the user to very precisely control what streams the callback will use.

Reply:
> For streams, for example, it allows the user to very precisely control what streams the callback will use.

Yep, same for device as well, especially when the callback fn is an imported function and users cannot easily change it. The current behavior should be sufficient to unblock RPC use cases. So I am OK to land this PR and modify it later if necessary.